src/caffe/layers/data_layer.cpp

```diff
@@ -14,7 +14,36 @@ namespace caffe {
 template <typename Dtype>
 DataLayer<Dtype>::DataLayer(const LayerParameter& param)
   : BasePrefetchingDataLayer<Dtype>(param),
-    reader_(param), pool_(3) {
+    reader_(param) {
+    // Check user override in param file
+    if (this->layer_param_.data_param().threads() == 0) {
+#ifndef CPU_ONLY
+        // In GPU mode, default to # CPU cores / # GPUs
+        if (Caffe::mode() == Caffe::GPU) {
+            int cuda_devices;
+            int cpu_count;
+            int prefetch_threads;
+            cpu_count = boost::thread::hardware_concurrency();
+            cudaGetDeviceCount(&cuda_devices);
+            DLOG(INFO) << "CPU threads: " << cpu_count
+                       << " GPU devices visible: " << cuda_devices;
+            prefetch_threads = cpu_count > 1 ? cpu_count / cuda_devices : 1;
```

Comment:
Would give 0 threads if cpu_count==2 and cuda_devices=4. Also would crash if cuda_devices=0.